File: csharp/src/AutomationFramework.Core/Support/TestValues.cs
using System.Globalization;
using Bogus;

namespace AutomationFramework.Core.Support;

/// <summary>
/// Generated values that are unique per run and reproducible on demand.
/// </summary>
/// <remarks>
/// The generator is seeded, and the seed is logged. Without that, a failure caused by a particular
/// generated value can never be reproduced: you get a red test, a re-run that passes, and no
/// explanation. Set <c>AF_DATA_SEED</c> to the logged value to replay a run's data exactly.
/// <para>
/// Generated identities use reserved test domains only. A real domain in test data eventually means
/// real email sent to a real person.
/// </para>
/// </remarks>
public static class TestValues
{
    private static readonly int Seed = ResolveSeed();

    private static readonly ThreadLocal<Faker> LocalFaker = new(() => new Faker
    {
        Random = new Randomizer(Seed),
    });

    static TestValues()
    {
        TestLog.Info(
            $"Test data seed: {Seed}. Set AF_DATA_SEED={Seed} to reproduce this run's generated data.");
    }

    /// <summary>Generator for the current thread. Seeded, so two threads produce the same stream independently.</summary>
    public static Faker Faker => LocalFaker.Value!;

    /// <summary>Unique, run-scoped email on a reserved domain that cannot deliver mail.</summary>
    public static string UniqueEmail()
        => string.Create(
            CultureInfo.InvariantCulture,
            $"af-{RunContext.RunId}-{RunContext.NextSequence()}@example.invalid");

    /// <summary>Unique, run-scoped name prefixed so a janitor job can find leftovers.</summary>
    public static string UniqueName(string prefix)
        => string.Create(
            CultureInfo.InvariantCulture,
            $"{prefix}-af-{RunContext.RunId}-{RunContext.NextSequence()}");

    /// <summary>
    /// An id that is syntactically valid for the target API but certain not to exist.
    /// </summary>
    /// <remarks>
    /// Shaped as a 26-character ULID because the demo API rejects anything else before it gets as far as
    /// looking the record up, which would test the wrong thing. A 404 test has to send an id the service
    /// accepts as well formed.
    /// </remarks>
    public static string WellFormedButMissingId() => "0AF00000000000000000000000";

    /// <summary>
    /// The seed for this run: <c>AF_DATA_SEED</c> when set, otherwise a fresh random one.
    /// </summary>
    /// <remarks>
    /// A fixed default would be worse than random. It would hide the whole class of bug where a test
    /// only passes for one particular generated value, because every run would use that same value.
    /// </remarks>
    private static int ResolveSeed()
    {
        string? fromEnvironment = Environment.GetEnvironmentVariable("AF_DATA_SEED");
        if (!string.IsNullOrWhiteSpace(fromEnvironment))
        {
            if (int.TryParse(
                    fromEnvironment.Trim(),
                    NumberStyles.Integer,
                    CultureInfo.InvariantCulture,
                    out int parsed))
            {
                return parsed;
            }

            throw new InvalidOperationException(
                $"AF_DATA_SEED must be a whole number, but was '{fromEnvironment}'.");
        }

        return Random.Shared.Next();
    }
}

File: java/src/main/java/com/company/automation/support/TestValues.java
package com.company.automation.support;

import java.util.Locale;
import java.util.Random;
import net.datafaker.Faker;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Generated values that are unique per run and reproducible on demand.
 *
 * <p>The faker is seeded, and the seed is logged. Without that, a failure caused by a particular
 * generated value can never be reproduced — you get a red test, a re-run that passes, and no
 * explanation. Set {@code AF_DATA_SEED} to the logged value to replay a run's data exactly.
 *
 * <p>Generated identities use reserved test domains only. A real domain in test data eventually
 * means real email sent to a real person.
 */
public final class TestValues {

  private static final Logger log = LoggerFactory.getLogger(TestValues.class);

  private static final long SEED = resolveSeed();
  private static final ThreadLocal<Faker> FAKER =
      ThreadLocal.withInitial(() -> new Faker(Locale.ENGLISH, new Random(SEED)));

  static {
    log.info(
        "Test data seed: {}. Set AF_DATA_SEED={} to reproduce this run's generated data.",
        SEED,
        SEED);
  }

  private TestValues() {}

  /**
   * The seed for this run: {@code AF_DATA_SEED} when set, otherwise a fresh random one.
   *
   * <p>A fixed default would be worse than random. It would hide the whole class of bug where a
   * test only passes for one particular generated value, because every run would use that same
   * value.
   */
  private static long resolveSeed() {
    String fromEnv = System.getenv("AF_DATA_SEED");
    if (fromEnv != null && !fromEnv.isBlank()) {
      try {
        return Long.parseLong(fromEnv.trim());
      } catch (NumberFormatException e) {
        throw new IllegalStateException(
            "AF_DATA_SEED must be a whole number, but was '" + fromEnv + "'.", e);
      }
    }
    return new java.util.Random().nextLong();
  }

  /** Faker for the current thread. Seeded, so two threads produce the same stream independently. */
  public static Faker faker() {
    return FAKER.get();
  }

  /** Unique, run-scoped email on a reserved domain that cannot deliver mail. */
  public static String uniqueEmail() {
    return "af-" + RunContext.runId() + "-" + RunContext.nextSequence() + "@example.invalid";
  }

  /** Unique, run-scoped name prefixed so a janitor job can find leftovers. */
  public static String uniqueName(String prefix) {
    return prefix + "-af-" + RunContext.runId() + "-" + RunContext.nextSequence();
  }

  /**
   * An id that is syntactically valid for the target API but certain not to exist.
   *
   * <p>Shaped as a 26-character ULID because the demo API rejects anything else before it gets as
   * far as looking the record up — which would test the wrong thing. A 404 test has to send an id
   * the service accepts as well formed.
   */
  public static String wellFormedButMissingId() {
    return "0AF00000000000000000000000";
  }
}
